x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+
+#### Normalization on read
+
+All endpoints that respond with a rule Saved Object, typed as `RuleResponse`, will perform **normalization on read**, which will transform legacy rules to the new schema **in memory** on read operation, before returning it as a response from an API endpoint.
+
+This means that the endpoints will always respond with the rules with the new schema, while the actual rule saved object might still be stored with the legacy schema in Elasticsearch, if it still has not been migrated-on-write.
+
+The **normalization on read** will be carried out by a new `normalizeRuleSourceSchemaOnRuleRead` normalization function. The `internalRuleToAPIResponse` method, which is used in our endpoints to convert a rule saved object as is stored in Elasticsearch to the `RuleResponse` type which is returned to the client, calls the `commonParamsCamelToSnake` methods to convert rule parameters that are common to all rule types to what's expected in `RuleResponse`. 
+
+Inside this method, we will use `normalizeRuleSourceSchemaOnRuleRead` to calculate the normalized values of `rule_source` and `immutable`.
+
+_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts)_
+
+```ts
+export const internalRuleToAPIResponse = (rule) => {
+  return {
+    ...commonParamsCamelToSnake(rule.params), // <--- rule params are converted here
+    ...typeSpecificCamelToSnake(rule.params),
+    // [... more object properties ...]
+  };
+};
+
+export const commonParamsCamelToSnake = (params: BaseRuleParams) => {
+  const { immutable, rule_source } = normalizeRuleSourceSchemaOnRuleRead(params);
+
+  return {
+    immutable,
+    rule_source,
+    // [... more object properties ...]
+  };
+};
+```
+
+And the `normalizeRuleSourceSchemaOnRuleRead` can be defined so:
+
+_Source: x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/prebuilt_rule_schema_migrations.ts_ (New file)
+
+```ts
+interface OnReadNormalizationResult {
+  immutable: IsRuleImmutable;
+  rule_source: RuleSource
+}
+
+/**
+ * To calculate `rule_source`:
+ * - Use `rule_source` field  if already exists in the internal rule object params.
+ * - If it does not exist, check the value `immutable` from the rules param:
+ *   - if it is `true`, create a new `rule_source` object, with `isCustomized` set to `false` and `repoNam` set to `elastic_prebuilt` (Use case of rule_source rules that have not been yet migrated-on-write)
+ *   - otherwise, the field should be `undefined` (use case of custom rules).
```

Comment:
This comment is not correct anymore, although I'm not sure if we should have it because the code is self-explanatory. Anyway:

```suggestion
 * To calculate `rule_source`:
 * - Use `rule_source` field  if already exists in the internal rule object params. This means the rule has been migreted-on-write.
 * - If `rule_source` does not exist (the rule hasn't been migrated yet), check the value of `immutable`:
 *   - if it is `true`, create a new `rule_source` object of type `external`, with `is_customized` set to `false`
 *   - if it is `false`, create a new `rule_source` object of type `internal`
```

Reply:
Relying on code, removing text explanation